### ISSUE TITLE
Proposal: Improve calendar date picker flyout editing

### ISSUE THREAD

**Issue by robloo** [microsoft/microsoft-ui-xaml#2255]
# Proposal: Improve calendar date picker flyout editing

We need to improve editing of the selected date in calendar date picker flyout. These improvements are necessary to speed up editing by the user.

This is related to but different from the improvements mentioned in #735.

## Summary
Speed up editing of the selected date in the calendar date picker by providing ComboBoxes for the month and year.

**Current Design**
![Untitled_1](https://user-images.githubusercontent.com/17993847/78815041-e1450280-798c-11ea-8ed8-38a9f835fb1e.png)

**Proposed Design**
![Untitled_2](https://user-images.githubusercontent.com/17993847/78815280-4a2c7a80-798d-11ea-8db4-1cb09d362288.png)

## Rationale
The current calendar date picker is slow to use for the user. However, it is visually more intuitive and therefore the preferred control in user interfaces.

## Scope

| Capability | Priority |
| :---------- | :------- |
| Add month selection ComboBox to the Flyout | Must |
| Add year selection ComboBox to the Flyout | Must |
| Make ComboBox visibility configurable (to keep current design for backwards compatibility) | Should |
| Make year selection ComboBox editable | Could |
| Make starting/ending year listed in the year selection ComboBox configurable | Could |

## Important Notes

 - The month selection ComboBox will have all 12 months listed from the current UI culture.
 - The year selection ComboBox will only have the last 10-20 years visible (to avoid making the list too long)
  - Making the year ComboBox editable would speed up year entry. However, there may be localization concerns doing this.

**Comment by mdtauk** [comment]
Isn't this why there is the DatePicker with the calendar flyout?

![image](https://user-images.githubusercontent.com/7389110/78834048-d8344f80-79e5-11ea-8e31-b939e73f6e35.png)

Maybe that should get an **IsEditable** property.  There was an issue about making those pickers better for keyboard use. #918 #905 #735

**Comment by robloo** [comment]
@mdtauk, I'm not sure I understand your comments entirely. I'm talking about the CalendarDatePicker which is different than the DatePicker. The CalendarDatePicker does look exactly as you pictured though. I did only screenshot and modify the relevant part -- the calendar flyout only.

> Maybe that should get an IsEditable property

I wouldn't suggest that for this case. I'm talking about only the flyout which is already the editor. Once the flyout is open the user is already editing, I'm only trying to make this experience faster. Now if we made the inline textbox area (not the flyout) editable like in #735 that could use the `IsEditable` property.

> There was an issue about making those pickers better for keyboard use. #918 #905 #735

Yes, I referred to #735 in the proposal but this is quite a bit different -- although the spirit is the same (to improve the editing experience).

**Comment by mdtauk** [comment]
@robloo I do not think your idea is a good solution for having an editable date picker.  The control is designed to be a fixed size, and the additional weight of combo boxes feels clunky.  I think keeping the control a pure calendar grid viewer is a good idea, but there are other ideas that achieve the same thing, fast way to pick a date, by handling it in the form control Date Picker or Calendar Picker.

**Comment by robloo** [comment]
@mdtauk, I respectfully disagree. There is also precedent for my suggestion. Microsoft has internally decided it is a valid solution to the problem:

![example](https://user-images.githubusercontent.com/17993847/78856091-fa789e00-79e2-11ea-9269-6fa323e467ed.gif)

This is taken from the developer dashboard. I don't know the web framework they used off the top of my head.

I see this as somewhat of an elegant solution as it maintains the same size of the current flyout; however, significantly speeds up selecting different months/years. No longer does the user have to click the Month/Year title to navigate to Year and Decade views (which is not only more time consuming but more complex I would argue). 

If someone likes to use more of a form-style control they can keep using the DatePicker or the CalendarView.

**Comment by robloo** [comment]
If it's the concept of nested flyouts you don't like, I can understand that as well (there is a framework bug for that...). It's also possible to rework the existing control to press the month and year **separately**. This could use existing designs instead of adding ComboBoxes. The downside is you loose the drop down symbol and ComboBox appearance so the user doesn't always know they can click that area (this has long been a design deficiency in this type of control).

Pressing the month title *April* goes directly to the month view (as it already does). 
![image](https://user-images.githubusercontent.com/17993847/78856491-28121700-79e4-11ea-9d08-cc122e2f27bb.png)

But pressing year title *2020* would go directly to the year view. This would prevent having to click the title twice.

![image](https://user-images.githubusercontent.com/17993847/78856526-40823180-79e4-11ea-842a-5f979fc04da5.png)

**Comment by mdtauk** [comment]
Making the Month and Year tap-able I think is a better option than ComboBoxes.

FluentUI has a property that does this. (formerly Fabric UI)
![image](https://user-images.githubusercontent.com/7389110/78857738-21a29000-7a22-11ea-9044-5320365f332a.png)

The new Edgium form control does it
![image](https://user-images.githubusercontent.com/7389110/78858004-04ba8c80-7a23-11ea-806c-8b66a6c4a7df.png)

----

You can sort of do what you want currently with the Xaml control
![image](https://user-images.githubusercontent.com/7389110/78858282-e012e480-7a23-11ea-84b5-1b39d10f8d66.png)

It just takes two taps to get to the year selection/decade view

**Comment by robloo** [comment]
> You can sort of do what you want currently with the Xaml control
> It just takes two taps to get to the year selection/decade view

Yes, I'm aware. It is slow and somewhat unintuitive to use (especially when entering a lot of dates in different months/years):
  1. There is no visual indication that the title can be pressed
  2. Pressing the title once goes to month view, twice goes to year view, and on older controls you could even press it a third time and go to a decade view I believe. This multiple pressing is not always ideal.
  3. The year, month, and day must all be set at once. There is no mechanism to modify only the year or month of the currently selected date. This means just changing the year requires re-selecting the month and day.

So it seems #3 is the main point I'm trying to improve. Make the day, month and year independently selectable. That could be done by improving the flyout UI as discussed above and also by making the box editable (#735).

**Comment by mdtauk** [comment]
> 3. The year, month, and day must all be set at once. There is no mechanism to modify only the year or month of the currently selected date. This means just changing the year requires re-selecting the month and day.
> 
> So it seems no. 3 is the main point I'm trying to improve. Make the day, month and year independently selectable. That could be done by improving the flyout UI as discussed above and also by making the box editable (#735).
>

![image](https://user-images.githubusercontent.com/7389110/78921266-8e527480-7a8c-11ea-9435-ac423fb0e9a9.png)
![image](https://user-images.githubusercontent.com/7389110/78921386-bc37b900-7a8c-11ea-90aa-72372c204abc.png)

Something like this would also make date picking simpler.

**Comment by mdtauk** [comment]
> I definitely think this is needed if for nothing more than to allow directly typing into the control as an alternative to selecting from a flyout. With a keyboard typing '7' [tab] '58' [tab] [up] to select 7:58pm would be a lot quicker and useful for desktop scenarios.

To make it work nicely, the whole control would need some tweaks.

![image](https://user-images.githubusercontent.com/7389110/78928046-dcb94080-7a97-11ea-92b9-adf9fda092d6.png)

**Comment by robloo** [comment]
Really, really, like the last set of illustrations. I'm not using a DatePicker because the calendar makes it easier to select dates in some cases.

**Comment by mdtauk** [comment]
![image](https://user-images.githubusercontent.com/7389110/78934644-32471a80-7aa3-11ea-910a-5f48fe33caab.png)
Here is one with a slightly cleaner header, similar to the Pivot

**Comment by robloo** [comment]
I've updated the proposal to capture the latest discussion. Thanks @mdtauk for the great illustrations and ideas.

**Comment by shaheedmalik** [comment]
> To make it work nicely, the whole control would need some tweaks.
> 
> ![image](https://user-images.githubusercontent.com/7389110/78928046-dcb94080-7a97-11ea-92b9-adf9fda092d6.png)

Can you make one with MM DD YY format?

**Comment by mdtauk** [comment]
> Can you make one with MM DD YY format?
>
![image](https://user-images.githubusercontent.com/7389110/79064064-8d446180-7c9d-11ea-8c23-6b635ffd9099.png)

**Comment by robloo** [comment]
Would still like to see a new design for faster use here. If Microsoft has no plans to do anything in this space though, please just close it.

@chigy
